### PR TITLE
feat(extension): allow specifying number of values for subcommand's argument

### PR DESCRIPTION
**Pull request by smallstepman**
# Description

I want to suggest adding `values` (_naming?_) parameter (`.subcommands.<SUBCMD>.args.<ARG>.values`) in the `extension.json` manifest schema, which specifies how many values an arg can take. The parameter can be either positive integer or string, and can represent:
- "unlimited" amount of values with a format like `"values": "unlimited"`
- a specific amount of values with a format like `"values": 4`, meaning that the argument must take exactly 4 values (developers can specify `0`, meaning "arg doesn't take values", useful for "pure flags", e.g. `cargo build --workspace`)
- a range of allowable values with a format like `"values": "2..5"`, meaning that the argument can take anywhere from 2 to 5 values

Replaces `"multiple": <bool>` parameter.

This came up in https://github.com/dfinity/dfx-extensions/pull/38

# How Has This Been Tested?

- [x] unit
- [ ] e2e (TODO - will add them in a separate PR to avoid blocking the beta release on Friday evening)

# Checklist:

- [x] The title of this PR complies with [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/).
- [ ] I have edited the CHANGELOG accordingly.
- [ ] ~I have made corresponding changes to the documentation.~

### CODE REVIEW

File: src/dfx-core/src/extension/manifest/extension.rs [hunk 1]
```diff
@@ -170,26 +230,75 @@ fn parse_test_file() {
...
+    let m: Result<ExtensionManifest, serde_json::Error> = dbg!(serde_json::from_str(f));
     assert!(m.is_ok());
 
-    let subcmds = m.unwrap().into_clap_commands().unwrap();
-    dbg!(&subcmds);
-    for s in &subcmds {
+    let mut subcmds = dbg!(m.unwrap().into_clap_commands().unwrap());
+
+    for s in &mut subcmds {
+        s.print_long_help().unwrap();
```
Comment (smallstepman) [inline src/dfx-core/src/extension/manifest/extension.rs:301]:
```console
running 1 test
About for config command. You're looking at the output of parsing test extension.json.

Usage: config [COMMAND]

Commands:
  create    Command line options for creating an SNS configuration.
  validate  Command line options for validating an SNS configuration.
  help      Print this message or the help of the given subcommand(s)

Options:
  -h, --help
          Print help
About for deploy command. You're looking at the output of parsing test extension.json.

Usage: deploy

Options:
  -h, --help
          Print help
About for download command. You're looking at the output of parsing test extension.json.

Usage: download [OPTIONS]

Options:
      --ic-commit <ic_commit>
          IC commit of SNS canister WASMs to download

      --wasms-dir <wasms_dir>
          Path to store downloaded SNS canister WASMs

  -h, --help
          Print help
About for import command. You're looking at the output of parsing test extension.json.

Usage: import [OPTIONS]

Options:
      --network-mapping <network_mapping>
          Networks to import canisters ids for.
            --network-mapping <network name in both places>
            --network-mapping <network name here>=<network name in project being imported>
          Examples:
            --network-mapping ic
            --network-mapping ic=mainnet

  -h, --help
          Print help
About for initialize-canister command. You're looking at the output of parsing test extension.json.

Usage: initialize-canister [canister_id]

Arguments:
  [canister_id]
          some arg that accepts multiple values separated by spaces

Options:
  -h, --help
          Print help
About for initialize-canisters command. You're looking at the output of parsing test extension.json.

Usage: initialize-canisters [canister_ids]...

Arguments:
  [canister_ids]...
          some arg that accepts multiple values separated by spaces

Options:
  -h, --help
          Print help
About for initialize-two-canisters command. You're looking at the output of parsing test extension.json.

Usage: initialize-two-canisters [canister_ids] [canister_ids]

Arguments:
  [canister_ids] [canister_ids]
          some arg that accepts multiple values separated by spaces

Options:
  -h, --help
          Print help
About for initialize-two-or-three-canisters command. You're looking at the output of parsing test extension.json.

Usage: initialize-two-or-three-canisters [canister_ids] [canister_ids]...

Arguments:
  [canister_ids] [canister_ids]...
          some arg that accepts multiple values separated by spaces

Options:
  -h, --help
          Print help
About for install command. You're looking at the output of parsing test extension.json.

Usage: install [OPTIONS]

Options:
      --accounts [<accounts>...]
          some arg that accepts multiple values separated by spaces

  -h, --help
          Print help
Usage: sns [COMMAND]

Commands:
  config                             About for config command. You're looking at the output of parsing test extension.json.
  deploy                             About for deploy command. You're looking at the output of parsing test extension.json.
  download                           About for download command. You're looking at the output of parsing test extension.json.
  import                             About for import command. You're looking at the output of parsing test extension.json.
  initialize-canister                About for initialize-canister command. You're looking at the output of parsing test extension.json.
  initialize-canisters               About for initialize-canisters command. You're looking at the output of parsing test extension.json.
  initialize-two-canisters           About for initialize-two-canisters command. You're looking at the output of parsing test extension.json.
  initialize-two-or-three-canisters  About for initialize-two-or-three-canisters command. You're looking at the output of parsing test extension.json.
  install                            About for install command. You're looking at the output of parsing test extension.json.
  help                               Print this message or the help of the given subcommand(s)

Options:
  -h, --help  Print help
test extension::manifest::extension::parse_test_file ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 37 filtered out; finished in 0.05s
```

**Review by smallstepman — COMMENTED** [review]
output of running that test

**Review by ericswanson-dfinity — APPROVED** [review]

**Comment by smallstepman** [comment]
@ericswanson-dfinity 
> failing tests

not sure how to move forward here, I see three options, could you help me pick one?:
1. update all `extensions.json` in dfx-extensions repo (e2e tests won't be able to pass there, but we don't have `:required` set for the workflow step in `e2e.yml`, therefore despite failing e2e, it will be possible to merge), and make a new release of both nns and sns extensions
2. patch extension.json in e2e tests in SDK repo (a script replacing all `"multiple": true` -> `"values": "unlimited"`), once next dfx beta release is published, I can do the replacement in dfx-extensions repo (and make a new release of the extensions), and a remove the patch from SDK repo
3. same as option number 2, but instead of patching, we can `skip` failing tests

**Comment by ericswanson-dfinity** [comment]
Another option: leave `multiple: bool` as a valid option in this PR (alias for values: unlimited).  Merge this PR, then release new extensions that use `values: unlimited`, then follow up here with a PR that removes `multiple: bool`.

**Review by ericswanson-dfinity — APPROVED** [review]

**Pull request merged**